pkg/storegateway/bucket_test.go

```diff
@@ -943,14 +943,44 @@ func createBlockFromHead(t testing.TB, dir string, head *tsdb.Head) ulid.ULID {
 	return ulid
 }
 
-// Very similar benchmark to ths: https://github.com/prometheus/prometheus/blob/1d1732bc25cc4b47f513cb98009a4eb91879f175/tsdb/querier_bench_test.go#L82,
-// but with postings results check when run as test.
 func benchmarkExpandedPostings(
-	t test.TB,
+	tb test.TB,
 	newTestBucketBlock func() *bucketBlock,
 	series int,
 ) {
-	ctx := context.Background()
+	ctx, cancel := context.WithCancel(context.Background())
+	tb.Cleanup(cancel)
+
+	seriesSelectionTestCases(tb, series, func(tb test.TB, testCase seriesSelectionTestCase) {
+		indexr := newBucketIndexReader(newTestBucketBlock())
+		indexrStats := newSafeQueryStats()
+
+		tb.ResetTimer()
+		for i := 0; i < tb.N(); i++ {
+			p, err := indexr.ExpandedPostings(ctx, testCase.matchers, indexrStats)
+
+			if err != nil {
+				tb.Fatal(err.Error())
+			}
+			if testCase.expectedSeriesLen != len(p) {
+				tb.Fatalf("expected %d postings but got %d", testCase.expectedSeriesLen, len(p))
+			}
+		}
+	})
+}
+
+type seriesSelectionTestCase struct {
+	name              string
+	matchers          []*labels.Matcher
+	expectedSeriesLen int
+}
+
+// Very similar benchmark to ths: https://github.com/prometheus/prometheus/blob/1d1732bc25cc4b47f513cb98009a4eb91879f175/tsdb/querier_bench_test.go#L82,
+func seriesSelectionTestCases(
+	t test.TB,
+	series int,
+	runTestCase func(tb test.TB, testCase seriesSelectionTestCase),
+) {
```

Comment:
I don't see the need for this to wrap another function. Can this be just:

```suggestion
// Very similar benchmark to ths: https://github.com/prometheus/prometheus/blob/1d1732bc25cc4b47f513cb98009a4eb91879f175/tsdb/querier_bench_test.go#L82,
func seriesSelectionTestCases(
	t test.TB,
	series int,
) []seriesSelectionTestCase {
```

And then let everyone do with those testcases whatever they want? I think that's more flexible.

Not a strong opinion, but I find it easier to reason on a test/benchmark that does:

```go
for _, c := range seriesSelectionTestCases(t, 500) {
    t.Run(c.name, func(t *testing.T) {
        doStuff(c.things)
    }
}
```

Because it's much clearer what is happening than on a function passed to a third-party function that might be doing something else.

Reply:
If we do that, we also don't have to stick to `test.TB` as we do now: each caller can use anything they want.

Reply:
I prefer your approach. Will merge this on green